joy_teleop/joy_teleop/joy_teleop.py

```diff
@@ -251,71 +229,167 @@ def run_topic(self, c, joy_state):
+
+            # Construct an example action Goal, and try to fill it in.  This is
+            # to give the user early feedback if their config can't possibly work.
+            goal = self.action_type.Goal()
+            for target, value in self.action_goal.items():
+                set_message_fields(goal, {target: value})
+
+        action_name = config['action_name']
+
+        self.action_client = ActionClient(node, self.action_type, action_name)
+
+    def run(self, node, joy_state):
+        if not self.check_buttons_and_axes(joy_state):
             return
-        target = msg
-        for i in ml[:-1]:
-            target = getattr(target, i)
-        setattr(target, ml[-1], value)
 
-    def get_interface_type(self, type_name, ext):
-        if type_name not in self.message_types:
+        self.set_ready(self.action_client.server_is_ready())
+
+        if not self.ready:
+            return
+
+        goal = self.action_type.Goal()
+        for target, value in self.action_goal.items():
+            set_message_fields(goal, {target: value})
+
+        self.action_client.send_goal_async(goal)
+
+
+class JoyTeleop(Node):
+    """
+    Generic joystick teleoperation node.
+
+    Will not start without configuration, has to be stored in 'teleop' parameter.
+    See config/joy_teleop.yaml for an example.
+    """
+
+    def __init__(self):
+        super().__init__('joy_teleop', allow_undeclared_parameters=True,
+                         automatically_declare_parameters_from_overrides=True)
+
+        self.commands = []
+
+        names = []
+
+        for name, config in self.retrieve_config().items():
+            if name in names:
+                self.get_logger().error("command '{}' was duplicated".format(name))
+                raise JoyTeleopException("command '{}' was duplicated".format(name))
+
             try:
-                package, interface, message = type_name.split('/')
-                mod = importlib.import_module(package + ext)
-                self.message_types[type_name] = getattr(mod, message)
-            except ValueError:
-                raise JoyTeleopException('message type format error')
-            except ImportError:
-                raise JoyTeleopException('module {} could not be loaded'.format(package))
-            except AttributeError:
-                raise JoyTeleopException(
-                    'message {} could not be loaded from module {}'.format(package, message))
-        return self.message_types[type_name]
-
-    def update_actions(self, evt=None):
-        for name, cmd in self.command_list.items():
-            if cmd['type'] != 'action':
-                continue
-            if cmd['action_name'] in self.offline_actions:
-                self.register_action(name, cmd)
+                interface_group = config['type']
+
+                if interface_group == 'topic':
+                    self.commands.append(JoyTeleopTopicCommand(name, config, self))
```

Comment:
I tried this, and it turns out to be exactly the same lines of code.  Since I don't expect the list of interface_groups to expand much over time, and since I find this more readable, I left it as-is.